### PR TITLE
Database (architecture + use) documentation

### PR DESCRIPTION
**Pull request by BioTheWolff**
Documents the database by describing its architecture, and how the components interact with it.

Checklist:

- [x] technologies documentation
- [x] backing up claims using benchmarks or proofs
- [x] architecture (ER) diagram

### CODE REVIEW

File: Architecture.md [hunk 1]
```diff
@@ -55,10 +63,62 @@
...
+
+Furthermore, since the projects will be deployed at a subdomain that is named after `https://{projectId}.user-app.paastech.cloud` we needed to hide the internal database structure that emerges with a serial id. Therefore, using a uuid prevents sharing sensitive data with everyone and prevents targeted attacks.
+
+
+The table `users` contains all the necessary information about each user. Upon user creation, an `email_nonce` is automatically created and a user account is only considered active once the email has been confirmed and the field `email_nonce` is null.
+Should the user wish to reset his password, the field `password_nonce` will contain a unique uuid allowing the user to reset his password.
```
Comment (sea-gull-diana) [inline Architecture.md:107]:
```suggestion
Should the user wish to reset their password, the PaaSTech API will insert a unique uuid into the field `password_nonce`. The user will then be able to change their password by providing the said uuid in the reset request.
```

File: Architecture.md [hunk 2]
```diff
@@ -55,10 +61,62 @@
...
+One user can have multiple projects or SSH keys.
+
+The `ssh_keys` specified by each user allow them to push their repository onto our git server. Each SSH key can have a name to make it easier to distinguish multiple keys, however, it is not required.
+A SSH key belongs to a user, so it will have the same permissions as the user on all his repositories at the moment.
```
Comment (sea-gull-diana) [inline Architecture.md:111]: The phrase doesn't seem very clear. I hope I interpreted it correctly. And since SSH is pronounced as "Es-es-ash", we need to use "an" as an article.
```suggestion
An SSH key belongs to a user and not to a particular project at the moment. Thus, it will allow its owner to access any of their repositories. If a user has more than one SSH key, it does not matter which one they use to access their projects.
```

File: Architecture.md [hunk 3]
```diff
@@ -55,10 +61,62 @@
...
+        timestamp created_at
+        timestamp updated_at "Nullable"
+        uuid user_id FK
+    }
+
+    USERS ||--o{ PROJECTS : manage
+    USERS ||--o{ SSH_KEYS : possess
+```
+
+As you can notice, every table has a uuid field as a primary key. Compared to `SERIAL` that is often used to identify a table, a uuid guarantees a better uniqueness across the database. A [Serial](https://www.postgresql.org/docs/current/datatype-numeric.html) may take up less space (4 bytes) than an [uuid](https://www.postgresql.org/docs/current/datatype-uuid.html) (16 bytes) but since it is a series of incrementing integers it offers information about the time of creation and makes it easier to guess the id whereas an uuid is generated randomly and nearly impossible to duplicate.
```
Comment (sea-gull-diana) [inline Architecture.md:101]:
```suggestion
As you can notice, every table has a uuid field as a primary key. Compared to `SERIAL` that is often used to identify a table, a uuid guarantees uniqueness across the whole database (and most likely across the whole universe). A [Serial](https://www.postgresql.org/docs/current/datatype-numeric.html) may take up less space (4 bytes) than a [uuid](https://www.postgresql.org/docs/current/datatype-uuid.html) (16 bytes) but since it is a series of incrementing integers it offers information about the time of creation and makes it easier to guess the id whereas a uuid is generated randomly and is nearly impossible to duplicate.
```

**Review by sea-gull-diana — CHANGES_REQUESTED** [review]
The database part seems good to me. Not sure what more we can add. Just a few corrections for now :)

File: Architecture.md [hunk 4]
```diff
@@ -98,17 +98,17 @@ erDiagram
...
 
 
 The table `users` contains all the necessary information about each user. Upon user creation, an `email_nonce` is automatically created and a user account is only considered active once the email has been confirmed and the field `email_nonce` is null.
-Should the user wish to reset his password, the field `password_nonce` will contain a unique uuid allowing the user to reset his password.
+Should the user wish to reset their password, a new UUID will be saved in the field `password_nonce`. To reset his password, the Client will need to provide said UUID together with the new password.
 One user can have multiple projects or SSH keys.
 
 The `ssh_keys` specified by each user allow them to push their repository onto our git server. Each SSH key can have a name to make it easier to distinguish multiple keys, however, it is not required.
-A SSH key belongs to a user, so it will have the same permissions as the user on all his repositories at the moment.
+An SSH key belongs to a user and not a project. Thus, it will allow the Client to access any of their repositories. Should a user possess more than one SSH key, every one will grant them access.
```
Comment (BioTheWolff) [inline Architecture.md:111]: `every one` => `each one`

File: Architecture.md [hunk 5]
```diff
@@ -32,7 +32,13 @@
...
+
+As market leader, PostgreSQL stands out for its performance and widespread use throughout the world. Not only is free and open-source, it allows also a larger set of extensions and features than other databases, like the support for more than 43 different [data types](https://www.postgresql.org/docs/current/datatype.html). Together with its flexibility and usage of [Multi-Version concurrency Control (MVCC)](https://www.postgresql.org/docs/current/glossary.html#GLOSSARY-CONCURRENCY) as well as our prior experiences made it more interesting that its major competitor [MySQL](https://www.mysql.com/).
+
+The API connects to the database using an ORM called [Prisma](https://www.prisma.io/). Prisma also automates the generation & migration of the database as well as creates the classes associated to each table. Being an ORM that is both performant and allows the reorganisation of SQL data into simple classes, it was an important addition to our technology stack. Its complete documentation, active development and strong support made it more interesting than its competitors like [TypeORM](https://typeorm.io/) or [Objection js](https://vincit.github.io/objection.js/).
```
Comment (BioTheWolff) [inline Architecture.md:41]:
```suggestion
The API connects to the database using an ORM called [Prisma](https://www.prisma.io/). Prisma automates the generation and migration of the database as well as creating the classes associated to each table. Being an ORM that is both performant and allows the reorganisation of SQL data into simple classes, it was an important addition to our technology stack. Its complete documentation, active development and strong support made it more interesting than its competitors like [TypeORM](https://typeorm.io/) or [Objection JS](https://vincit.github.io/objection.js/).
```

File: Architecture.md [hunk 6]
```diff
@@ -55,10 +61,62 @@
...
+        uuid user_id FK
+    }
+
+    USERS ||--o{ PROJECTS : manage
+    USERS ||--o{ SSH_KEYS : possess
+```
+
+As you can notice, every table has a UUID field as a primary key. Compared to `SERIAL` that is often used to identify a table, a UUID guarantees a better uniqueness across the whole database. A [Serial](https://www.postgresql.org/docs/current/datatype-numeric.html) may take up less space (4 bytes) than a [UUID](https://www.postgresql.org/docs/current/datatype-uuid.html) (16 bytes) but since it is a series of incrementing integers it offers information about the time of creation and makes it easier to guess the id whereas a UUID is generated randomly and is nearly impossible to duplicate.
+
+Furthermore, since the projects will be deployed at a subdomain that is named after `https://{projectId}.user-app.paastech.cloud` we needed to hide the internal database structure that emerges with a serial id. Therefore, using a uuid prevents sharing sensitive data with everyone and prevents targeted attacks.
```
Comment (BioTheWolff) [inline Architecture.md:103]: The domain name should not contain the protocol.

```suggestion
Furthermore, since the projects will be deployed at a subdomain that is named after `{projectId}.user-app.paastech.cloud` we needed to hide the internal database structure that emerges with a serial id. Therefore, using a uuid prevents sharing sensitive data with everyone and prevents targeted attacks.
```

File: Architecture.md [hunk 7]
```diff
@@ -55,10 +61,62 @@
...
+        timestamp created_at
+        timestamp updated_at "Nullable"
+        uuid user_id FK
+    }
+
+    USERS ||--o{ PROJECTS : manage
+    USERS ||--o{ SSH_KEYS : possess
+```
+
+As you can notice, every table has a UUID field as a primary key. Compared to `SERIAL` that is often used to identify a table, a UUID guarantees a better uniqueness across the whole database. A [Serial](https://www.postgresql.org/docs/current/datatype-numeric.html) may take up less space (4 bytes) than a [UUID](https://www.postgresql.org/docs/current/datatype-uuid.html) (16 bytes) but since it is a series of incrementing integers it offers information about the time of creation and makes it easier to guess the id whereas a UUID is generated randomly and is nearly impossible to duplicate.
```
Comment (BioTheWolff) [inline Architecture.md:101]: If wanted, you could add another paragraph going into details of the difference between serials and uuids:

- serials are limited to 2147483647 (roughly 2 billion)
- serials *will* eventually clog up and be redundant
- we use the UUIDv4 standard, so the uuids are randomly generated
- there is a one-in-a-billion chance to find a duplicate uuid
- there are roughly 103 trillion UUIDv4s

File: Architecture.md [hunk 8]
```diff
@@ -55,10 +61,62 @@
...
+
+Furthermore, since the projects will be deployed at a subdomain that is named after `https://{projectId}.user-app.paastech.cloud` we needed to hide the internal database structure that emerges with a serial id. Therefore, using a uuid prevents sharing sensitive data with everyone and prevents targeted attacks.
+
+
+The table `users` contains all the necessary information about each user. Upon user creation, an `email_nonce` is automatically created and a user account is only considered active once the email has been confirmed and the field `email_nonce` is null.
+Should the user wish to reset their password, a new UUID will be saved in the field `password_nonce`. To reset his password, the Client will need to provide said UUID together with the new password.
```
Comment (BioTheWolff) [inline Architecture.md:107]: Once the password has been reset, the nonce is reset to null. Another attempt to reset the password overwrites the old nonce.

**Review by BioTheWolff — COMMENTED** [review]
@MurielParaire

File: Architecture.md [hunk 9]
```diff
@@ -55,10 +61,66 @@
...
+        uuid user_id FK
+    }
+
+    USERS ||--o{ PROJECTS : manage
+    USERS ||--o{ SSH_KEYS : possess
+```
+
+As you can notice, every table has a UUID field as a primary key. Compared to [Serial](https://www.postgresql.org/docs/current/datatype-numeric.html) that is often used to identify a table, a UUID guarantees a better uniqueness across the whole database. 
+To generate a [UUID](https://www.postgresql.org/docs/current/datatype-uuid.html), we are using the UUIDv4 standard, which generates each UUID randomly. Thus, there are roughly 103 trillion UUIDv4s, lowering the chance of finding a duplicate UUID to one-in-a-billion. 
+Compared to them, Serials are limited to roughly 2 billion (2147483647) and will eventually clog up an dbe reduntant.
```
Comment (BioTheWolff) [inline Architecture.md:103]:
```suggestion
In comparison, Serials are limited to roughly 2 billion (2147483647) and will eventually clog up the column and be reduntant. As the column is a primary key, this results in the table being full and creates an inability to insert new rows. 
```

File: Architecture.md [hunk 10]
```diff
@@ -55,10 +61,66 @@
...
+    }
+
+    USERS ||--o{ PROJECTS : manage
+    USERS ||--o{ SSH_KEYS : possess
+```
+
+As you can notice, every table has a UUID field as a primary key. Compared to [Serial](https://www.postgresql.org/docs/current/datatype-numeric.html) that is often used to identify a table, a UUID guarantees a better uniqueness across the whole database. 
+To generate a [UUID](https://www.postgresql.org/docs/current/datatype-uuid.html), we are using the UUIDv4 standard, which generates each UUID randomly. Thus, there are roughly 103 trillion UUIDv4s, lowering the chance of finding a duplicate UUID to one-in-a-billion. 
+Compared to them, Serials are limited to roughly 2 billion (2147483647) and will eventually clog up an dbe reduntant. 
+However, they take up less space (4 bytes) than a UUID (16 bytes) but since it is a series of incrementing integers it offers information about the time of creation and makes it easier to guess the id whereas a UUID is generated randomly and is nearly impossible to duplicate.
```
Comment (BioTheWolff) [inline Architecture.md:104]: Sentence is too long. Please shorten it and get to the point

Suggested (can be copied as-is):
```suggestion
However, even though serials take up less space (4 bytes) than a UUID (16 bytes), the column becomes a series of incrementing integers and offers information about the time of creation. This makes the table prone to timestamp-guessing. UUIDs are generated randomly and impossible to guess, completely removing this attack pattern.
```

File: Architecture.md [hunk 11]
```diff
@@ -55,10 +61,66 @@
...
+
+
+Furthermore, since the projects will be deployed at a subdomain that is named after `{projectId}.user-app.paastech.cloud` we needed to hide the internal database structure that emerges with a serial id. Therefore, using a uuid prevents sharing sensitive data with everyone and prevents targeted attacks.
+
+
+The table `users` contains all the necessary information about each user. Upon user creation, an `email_nonce` is automatically created and a user account is only considered active once the email has been confirmed and the field `email_nonce` is null.
```
Comment (BioTheWolff) [inline Architecture.md:110]: replace `user` with `User` (User = user account)

File: Architecture.md [hunk 12]
```diff
@@ -55,10 +61,66 @@
...
+Should the user wish to reset their password, a new UUID will be saved in the field `password_nonce`. To reset his password, the Client will need to provide said UUID together with the new password. Should another request to change the password been made before the password was reset, a new UUID will be generated and the old one will be overwritten. Once the password has been reset, `password_nonce` will be set to null.
```
Comment (BioTheWolff) [inline Architecture.md:111]: Defines what nonces are. Replace further occurences of "UUID" with "nonce" when possible, so the reader can follow along

File: Architecture.md [hunk 13]
```diff
@@ -55,10 +61,66 @@
...
+One user can have multiple projects or SSH keys.
+
+The `ssh_keys` specified by each user allow them to push their repository onto our git server. Each SSH key can have a name to make it easier to distinguish multiple keys, however, it is not required.
+An SSH key belongs to a user and not a project. Thus, it will allow the Client to access any of their repositories. Should a user possess more than one SSH key, each one will grant them access.
+
+
+The `projects` table describes a project. Its field `config` contains all the environmental variables of the user, like database authentication. since the configuration and necessary variables change for every project we decided to store it as a flexible json field. We decided to use a jsonb field that stores the json data in binary form, allowing for better performances than a simple json field.
```
Comment (BioTheWolff) [inline Architecture.md:118]: `project` should become `Project`.

```suggestion
The `projects` table describes a Project. Its field `config` contains all the environmental variables of said Project, like database URL. Since the configuration changes for every project, we decided to store it as a flexible json field. We decided to use a JSONB field that stores the JSON data in binary form, allowing for better performances than a simple json field.
```

File: Architecture.md [hunk 14]
```diff
@@ -60,7 +60,14 @@ It is used because it discovers when containers are started/stopped and can dyna
 
 #### Database
 
-**_TODO: [CLIENT] fill for the unified database schema_**
+The requirement for an ACID (Atomicity, Consistency, Isolation, and Durability) database led us to choose a relational database. For our application there wasn't a need to have a multitude of adaptable fields and the benefits of having consistent data with relationships between each other as well as having structured data were a more important factor.
```
Comment (sea-gull-diana) [inline Architecture.md:63]:
```suggestion
The requirement for an ACID (Atomicity, Consistency, Isolation, and Durability) database led us to choose a relational database. For our application there wasn't a need to have a multitude of adaptable fields and the benefits of having consistent and structured data with relationships between different objects were a more important factor.
```

File: Architecture.md [hunk 15]
```diff
@@ -60,7 +60,14 @@ It is used because it discovers when containers are started/stopped and can dyna
...
+
+The only flexible field is the environmental variables that are stored for each Project, but since our chosen database, [PostgreSQL](https://www.postgresql.org/), supports the [JsonB](https://www.postgresql.org/docs/9.5/datatype-json.html) type, this could easily be implemented.
+
+As market leader, PostgreSQL stands out for its performance and widespread use throughout the world. Not only is it free and open-source, it also allows also a larger set of extensions and features than other databases, like the support for more than 43 different [data types](https://www.postgresql.org/docs/current/datatype.html). Together with its flexibility and usage of [Multi-Version concurrency Control (MVCC)](https://www.postgresql.org/docs/current/glossary.html#GLOSSARY-CONCURRENCY) as well as our prior experiences made it more interesting that its major competitor [MySQL](https://www.mysql.com/).
```
Comment (sea-gull-diana) [inline Architecture.md:67]:
```suggestion
As market leader, PostgreSQL stands out for its performance and widespread use throughout the world. Not only is it free and open-source, it also offers a larger set of extensions and features than other databases, like the support for more than 43 different [data types](https://www.postgresql.org/docs/current/datatype.html). All of this together with its flexibility, the usage of [Multi-Version concurrency Control (MVCC)](https://www.postgresql.org/docs/current/glossary.html#GLOSSARY-CONCURRENCY) and our prior experiences made it a more interesting option that its major competitor [MySQL](https://www.mysql.com/).
```

File: Architecture.md [hunk 16]
```diff
@@ -182,10 +189,65 @@ As described previously, the API manages most of the user interactions and redir
 
 ### Database architecture
 
-**_TODO: [CLIENT] describe the database architecture, as referenced in the MCD in the README_**
+As the database is unified and serves as a focal point of the application, everything is contained in the same logical database.
```
Comment (sea-gull-diana) [inline Architecture.md:192]: I don't exactly understand the meaning here. Why the database is logical? 
```suggestion
As the database is unified and serves as a focal point of the application, it contains all of the data necessary for the functioning of PaaSTech's different components.
```

File: Architecture.md [hunk 17]
```diff
@@ -182,10 +189,65 @@ As described previously, the API manages most of the user interactions and redir
...
+        uuid user_id FK
+    }
+
+    USERS ||--o{ PROJECTS : manage
+    USERS ||--o{ SSH_KEYS : possess
+```
+
+As you can notice, every table has a UUID field as a primary key. Compared to [Serial](https://www.postgresql.org/docs/current/datatype-numeric.html) that is often used to identify a table, a UUID guarantees a better uniqueness across the whole database. 
+To generate a [UUID](https://www.postgresql.org/docs/current/datatype-uuid.html), we are using the UUIDv4 standard, which generates each UUID randomly. Thus, there are roughly 103 trillion UUIDv4s, lowering the chance of finding a duplicate UUID to one-in-a-billion. 
+In comparison, Serials are limited to roughly 2 billion (2147483647) and will eventually clog up the column and be reduntant. As the column is a primary key, this results in the table being full and creates an inability to insert new rows.
```
Comment (sea-gull-diana) [inline Architecture.md:231]:
```suggestion
In comparison, Serials are limited to roughly 2 billion (2147483647) and will eventually clog up the column and become redundant. As the column is a primary key, this results in the table being full and creates an inability to insert new rows. 
```

File: Architecture.md [hunk 18]
```diff
@@ -182,10 +189,65 @@ As described previously, the API manages most of the user interactions and redir
...
+    }
+
+    USERS ||--o{ PROJECTS : manage
+    USERS ||--o{ SSH_KEYS : possess
+```
+
+As you can notice, every table has a UUID field as a primary key. Compared to [Serial](https://www.postgresql.org/docs/current/datatype-numeric.html) that is often used to identify a table, a UUID guarantees a better uniqueness across the whole database. 
+To generate a [UUID](https://www.postgresql.org/docs/current/datatype-uuid.html), we are using the UUIDv4 standard, which generates each UUID randomly. Thus, there are roughly 103 trillion UUIDv4s, lowering the chance of finding a duplicate UUID to one-in-a-billion. 
+In comparison, Serials are limited to roughly 2 billion (2147483647) and will eventually clog up the column and be reduntant. As the column is a primary key, this results in the table being full and creates an inability to insert new rows. 
+However, even though serials take up less space (4 bytes) than a UUID (16 bytes), the column becomes a series of incrementing integers and offers information about the time of creation. This makes the table prone to timestamp-guessing. UUIDs are generated randomly and impossible to guess, completely removing this attack pattern.
```
Comment (sea-gull-diana) [inline Architecture.md:232]: I don't think the above paragraphs calls for "However" at the beginning of this one.
```suggestion
Even though serials take up less space (4 bytes) than a UUID (16 bytes), the column becomes a series of incrementing integers and offers information about the time of creation. This makes the table prone to timestamp-guessing. UUIDs are generated randomly and are impossible to guess, completely removing this attack pattern.
```

File: Architecture.md [hunk 19]
```diff
@@ -182,10 +189,65 @@ As described previously, the API manages most of the user interactions and redir
...
+
+
+Furthermore, since the projects will be deployed at a subdomain that is named after `{projectId}.user-app.paastech.cloud` we needed to hide the internal database structure that emerges with a serial id. Therefore, using a uuid prevents sharing sensitive data with everyone and prevents targeted attacks.
```
Comment (sea-gull-diana) [inline Architecture.md:235]:
```suggestion
Furthermore, since the projects will be deployed at a subdomain that is named after `{projectId}.user-app.paastech.cloud` we needed to hide the internal database structure that emerges with a serial id. Therefore, using a UUID prevents sharing sensitive data and targeted attacks.
```

File: Architecture.md [hunk 20]
```diff
@@ -182,10 +189,65 @@ As described previously, the API manages most of the user interactions and redir
...
+
+
+The table `users` contains all the necessary information about each User. Upon User creation, an `email_nonce` is automatically created and a Client account is only considered active once the email has been confirmed and the field `email_nonce` is null.
+A nonce is a random value that is unique and serves for identification. Should the User wish to reset their password, a new UUID will be saved in the field `password_nonce`. To reset his password, the Client will need to provide said nonce together with the new password. Should another request to change the password been made before the password was reset, a new nonce will be generated and the old one will be overwritten. Once the password has been reset, `password_nonce` will be set to null.
```
Comment (sea-gull-diana) [inline Architecture.md:239]:
```suggestion
A nonce is a random value that is unique and serves for identification. Should the User wish to reset their password, a new UUID will be saved in the field `password_nonce`. To reset his password, the Client will need to provide said nonce together with the new password. Should another request to change the password be made before the password was reset, a new nonce will be generated and the old one will be overwritten. Once the password has been reset, `password_nonce` will be set to null. 
```

**Review by sea-gull-diana — CHANGES_REQUESTED** [review]
A couple suggestions :)

File: Architecture.md (hunk 16 above)
Comment (MurielParaire) [reply Architecture.md:192]: If you prefer I can remove the logical, however I would rephrase it more like this : 
```
As the database is unified and serves as a focal point of the application, all the data is contained in the same database.
```
Personally I think it is pretty clear that it is all the data for PaaSTech so imo no need to explain that

Comment (BioTheWolff) [reply Architecture.md:192]: The database is logical, as opposed to physical. The physical database is the instance of the RDBMS, whereas the logical database is a database in that instance

File: Architecture.md [hunk 21]
```diff
@@ -182,10 +189,65 @@ As described previously, the API manages most of the user interactions and redir
...
+In comparison, Serials are limited to roughly 2 billion (2147483647) and will eventually clog up the column and be reduntant. As the column is a primary key, this results in the table being full and creates an inability to insert new rows. 
+However, even though serials take up less space (4 bytes) than a UUID (16 bytes), the column becomes a series of incrementing integers and offers information about the time of creation. This makes the table prone to timestamp-guessing. UUIDs are generated randomly and impossible to guess, completely removing this attack pattern.
+
+
+Furthermore, since the projects will be deployed at a subdomain that is named after `{projectId}.user-app.paastech.cloud` we needed to hide the internal database structure that emerges with a serial id. Therefore, using a uuid prevents sharing sensitive data with everyone and prevents targeted attacks.
+
+
+The table `users` contains all the necessary information about each User. Upon User creation, an `email_nonce` is automatically created and a Client account is only considered active once the email has been confirmed and the field `email_nonce` is null.
```
Comment (Mathias-Boulay) [inline Architecture.md:238]: I'm not sure about this one, and I'm starting to be tired.

```suggestion
The table `users` contains all the necessary information about each User. Upon User creation, an `email_nonce` is automatically created. A User is only considered active once the email has been confirmed, signified by the field `email_nonce` being set to null.
```

**Review by sea-gull-diana — APPROVED** [review]

File: Architecture.md (hunk 20 above)
Comment (Mathias-Boulay) [inline Architecture.md]: I also took the liberty to span the text across multiple lines, since my screen is only a 16:9 one, not a 9000:1 one.

```suggestion

A nonce is a random value that is unique and serves for identification. 
Should the User wish to reset their password, a new UUID will be saved in the field `password_nonce`.
The Client will need to provide said nonce together with the new password to finish the procedure.

In the case of multiple reset requests, the field `password_nonce` gets overwritten each time. This ensures only the last request stays valid.
Once the password has been reset, `password_nonce` will be set to null. 
```

File: Architecture.md [hunk 22]
```diff
@@ -182,10 +189,65 @@ As described previously, the API manages most of the user interactions and redir
...
+A nonce is a random value that is unique and serves for identification. Should the User wish to reset their password, a new UUID will be saved in the field `password_nonce`. To reset his password, the Client will need to provide said nonce together with the new password. Should another request to change the password been made before the password was reset, a new nonce will be generated and the old one will be overwritten. Once the password has been reset, `password_nonce` will be set to null. 
+One User can have multiple projects or SSH keys.
+
+The `ssh_keys` specified by each User allow them to push their repository onto our git server. Each SSH key can have a name to make it easier to distinguish multiple keys, however, it is not required.
+An SSH key belongs to a Client and not a Project. Thus, it will allow the Client to access any of their repositories. Should a User possess more than one SSH key, each one will grant them access.
```
Comment (Mathias-Boulay) [inline Architecture.md:243]:
```suggestion
An SSH key belongs to a Client and not a Project. Thus, the Client can access all of their repositories via any of their keys.
```

File: Architecture.md [hunk 23]
```diff
@@ -182,10 +189,65 @@ As described previously, the API manages most of the user interactions and redir
...
+
+
+The `projects` table describes a Project. Its field `config` contains all the environmental variables of said Project, like a database URL. Since the configuration changes for every Project, we decided to store it as a flexible json field. We decided to use a JSONB field that stores the JSON data in binary form, allowing for better performances than a simple json field.
```
Comment (Mathias-Boulay) [inline Architecture.md:246]: I also spanned it across multiple lines (2 actually, but that's a 100% more !)

Wait while I'm at it, where do you back your performance claims about JSONB compared to JSON ?
```suggestion
The `projects` table describes a Project. Its `config` field contains all the environment variables of said Project, like a database URL. 
Since the configuration changes for every Project, we decided to store it as a flexible JSON field. We decided to use a JSONB field that stores the JSON data in binary form, allowing for better performance than a simple JSON field.
```

**Review by Mathias-Boulay — CHANGES_REQUESTED** [review]
*Database schema not taken into account during the review.*
It felt quite hard to read at times, but the ideas are there. I took the liberty of suggesting changes to streamline the reader's experience.

File: Architecture.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (MurielParaire) [reply Architecture.md:192]: Yeah it is mostly used in optics, however it does have a second meaning which is "the point at which all elements or aspects converge; center of activity or attention:"

File: Architecture.md [hunk 24]
```diff
@@ -60,7 +60,13 @@ It is used because it discovers when containers are started/stopped and can dyna
 
 #### Database
 
-**_TODO: [CLIENT] fill for the unified database schema_**
+The requirement for an ACID (Atomicity, Consistency, Isolation, and Durability) database led us to choose a relational database. For our application there wasn't a need to have a multitude of adaptable fields and the benefits of having consistent and structured data with relationships between different objects were more important factors.
+
+The only flexible field is the environment variables that are stored for each Project, but since our chosen database, [PostgreSQL](https://www.postgresql.org/), supports the [JsonB](https://www.postgresql.org/docs/9.5/datatype-json.html) type, this was easy to implement.
```
Comment (Mathias-Boulay) [inline Architecture.md:65]:
```suggestion
The only flexible field is the environment variables that are stored for each Project, but since our chosen database, [PostgreSQL](https://www.postgresql.org/), supports the [JSONB](https://www.postgresql.org/docs/9.5/datatype-json.html) type, this was easy to implement.
```

**Review by Mathias-Boulay — CHANGES_REQUESTED** [review]
I missed this one in the first pass it seems. Only need to change this, then we're good !

File: Architecture.md [hunk 25]
```diff
@@ -182,10 +188,68 @@ As described previously, the API manages most of the user interactions and redir
...
+The `ssh_keys` specified by each User allow them to push their repository onto our git server. Each SSH key can have a name to make it easier to distinguish multiple keys, however, it is not required.
+An SSH key belongs to a Client and not a Project. Thus, the Client can access all of their repositories via any of their keys.
+
+
+The `projects` table describes a Project. Its `config` field contains all the environment variables of said Project, like a database URL. 
+Since the configuration changes for every Project, we decided to store it as a flexible JSON field. We decided to use a JSONB field that stores the JSON data in binary form, allowing for better performance than a simple JSON field.
 
 ### Detailed specification
 
+
```
Comment (Martin-Moreira-de-jesus) [inline Architecture.md:252]: what's the point of this line break ?

File: Architecture.md [hunk 26]
```diff
@@ -71,7 +77,7 @@ It is used because it discovers when containers are started/stopped and can dyna
 ### Definitions
 
 - the Service refers to PaaSTech as a whole;
-- a Client is a user account created by an end user against the Service;
+- a Client (also referred to as a User) is a user account created by an end user against the Service;
 - a Project is a materialisation of a Git repository, created by a Client using either the web frontend or the CLI. A Project can be deployed by the Client by pushing its code to the Service.
```
Comment (Martin-Moreira-de-jesus) [inline Architecture.md:81]: What does materialization mean in this context ?

**Review by Mathias-Boulay — APPROVED** [review]
Good work :+1:

File: Architecture.md (hunk 25 above)
Comment (BioTheWolff) [reply Architecture.md:252]: No point, this will be solved in a few seconds

File: Architecture.md (hunk 26 above)
Comment (BioTheWolff) [reply Architecture.md:81]: This is a requested thing, as User has been used a lot in place of Client.

Comment (Martin-Moreira-de-jesus) [reply Architecture.md:81]: I meant the sentence "a Project is a materialisation of a Git repository"

File: Architecture.md [hunk 27]
```diff
@@ -60,7 +60,13 @@ It is used because it discovers when containers are started/stopped and can dyna
 
 #### Database
 
-**_TODO: [CLIENT] fill for the unified database schema_**
+The requirement for an ACID (Atomicity, Consistency, Isolation, and Durability) database led us to choose a relational database. For our application there wasn't a need to have a multitude of adaptable fields and the benefits of having consistent and structured data with relationships between different objects were more important factors.
```
Comment (Martin-Moreira-de-jesus) [inline Architecture.md:63]: I think you should emphasize more on transactions, and how important they are to our use case.

**Review by Martin-Moreira-de-jesus — CHANGES_REQUESTED** [review]

File: Architecture.md (hunk 27 above)
Comment (BioTheWolff) [reply Architecture.md:63]: Will do

**Review by Martin-Moreira-de-jesus — APPROVED** [review]
